R/utils_render_grob.R

```diff
@@ -0,0 +1,1206 @@
+
+  cell_dims <- get_cell_dims(main_table_components, grob_ncols)
+
+  if(!is.null(header)){
+    grob_nrows <- grob_nrows + length(header)
+    header_dims <- get_cell_dims(header, 1)
+    cell_dims$heights <- c(header_dims$heights, cell_dims$heights)
+  }
+
+  if(!is.null(footer)){
+    grob_nrows <- grob_nrows + length(footer)
+    footer_dims <- get_cell_dims(footer, length(footer))
+    cell_dims$heights <- c(cell_dims$heights, footer_dims$heights)
+  }
+
+  if(!is.null(source_notes)){
+    grob_nrows <- grob_nrows + length(source_notes)
+    source_notes_dims <- get_cell_dims(source_notes, length(source_notes))
+    cell_dims$heights <- c(cell_dims$heights, source_notes_dims$heights)
+  }
+
+  col_widths <- cell_width_units(cell_dims$widths)
+  row_heights <- cell_height_units(cell_dims$heights)
+
+  grob_table <- gtable(widths = col_widths, heights = row_heights)
+
+  if(!is.null(header)){
+    for(i in seq_along(header)){
+      grob_table <- gg_add_table_row(grob_table, header[[i]], row = i , clip = "off")
+    }
+
+    offset <- i
+
+  }else{
+    offset <- 0
+  }
+
+
+  for(i in seq_along(main_table_components)){
+    grob_table <- gg_add_table_row(grob_table, main_table_components[[i]], row = i + offset)
+  }
+
+  offset <- i + offset
+
+  if(!is.null(footer)){
+    for(i in seq_along(footer)){
+      grob_table <- gg_add_table_row(grob_table, footer[[i]], row = i + offset)
+    }
+  }
+
+  offset <- i + offset
+
+
+  if(!is.null(source_notes)){
+    for(k in seq_along(source_notes)){
+      grob_table <- gg_add_table_row(grob_table, source_notes[[k]], row = k + offset)
+    }
+  }
+
+  structure(
+    grob_table,
+    class = c("gt_grob", class(grob_table))
+  )
+
+}
+
+#' @importFrom grid grid.newpage grid.draw
+#' @export
+print.gt_grob <- function(x,  ...) {
+  grid.newpage()
+  grid.draw(x)
+}
+
+gg_add_table_row <- function(g, cells, row ,...){
+
+  for(cell in cells){
+    g <- gg_table_add_grob(
+      g,
+      cell,
+      col = attr(cell, ".col"),
+      row = row,
+      span_cols = attr(cell, ".cell_col_span"),
+      span_rows = attr(cell, ".cell_row_span"),
+      ...
+    )
+  }
+
+  g
+
+}
+
+#' @importFrom gtable gtable_add_grob
+gg_table_add_grob <- function(g, grob, col, row, span_cols = 1, span_rows = 1, ...){
+  gtable_add_grob(
+    x = g,
+    grobs = grob,
+    l = col,  r = col + (span_cols - 1),
+    t = row,  b = row + (span_rows - 1),
+    ...
+  )
+}
+
+get_cell_dims <- function(x, ncol = 1){
+
+  heights <- rep(0,length(x))
+  widths <- rep(0, ncol)
+
+  for(r in seq_along(x)){
+    row_height <- heights[[r]]
+
+    for(cell in x[[r]]){
+
+      column <- attr(cell, ".col")
+      column_span <- attr(cell, ".cell_col_span")
+
+      font_size <- attr(cell, ".text_size")
+      cell_rows <- attr(cell, ".text_rows")
+      cell_nchar <- attr(cell, ".text_cols")
+
+      ## guestimation to what the height/width is
+      height <- font_size/9 * cell_rows
+      width <-  max(cell_nchar,1)/column_span
```

Comment:
Is there a reason why you are guessing here and not defering this to when the gross are constructed and you can ask grid what their dimensions are?

Reply:
None other than I wasn't sure how to do it. If we can easily determine the size of the grobs going in, then I would happily prefer that to trying to calculate it myself